ntoskrnl/fsrtl/fastio.c

```diff
@@ -1563,55 +1670,36 @@ FsRtlAcquireFileExclusiveCommon(IN PFILE_OBJECT FileObject,
     PEXTENDED_DRIVER_EXTENSION DriverExtension;
     PFS_FILTER_CALLBACKS FilterCallbacks;
+    FS_FILTER_PARAMETERS FilterParameters;
+    PVOID CompletionContext;
+    NTSTATUS Status = STATUS_SUCCESS;
 
     /* Get Device Object and Fast Calls */
     FcbHeader = (PFSRTL_COMMON_FCB_HEADER)FileObject->FsContext;
     DeviceObject = IoGetRelatedDeviceObject(FileObject);
 
-    /* Get master FsRtl lock */
+    /* Enter the master file system */
     FsRtlEnterFileSystem();
 
     DriverExtension = (PEXTENDED_DRIVER_EXTENSION)DeviceObject->DriverObject->DriverExtension;
     FilterCallbacks = DriverExtension->FsFilterCallbacks;
 
-    /* Check if Filter Cllbacks are supported */
-    if (FilterCallbacks && FilterCallbacks->PreAcquireForSectionSynchronization)
-    {
-        NTSTATUS Status;
-        PVOID CompletionContext;
-
-        FS_FILTER_CALLBACK_DATA CbData;
-
-        RtlZeroMemory(&CbData, sizeof(CbData));
-
-        CbData.SizeOfFsFilterCallbackData = sizeof(CbData);
-        CbData.Operation = FS_FILTER_ACQUIRE_FOR_SECTION_SYNCHRONIZATION;
-        CbData.DeviceObject = DeviceObject;
-        CbData.FileObject = FileObject;
-        CbData.Parameters.AcquireForSectionSynchronization.PageProtection = Reserved;
-        CbData.Parameters.AcquireForSectionSynchronization.SyncType = SyncType;
+    RtlZeroMemory(&FilterParameters, sizeof(FilterParameters));
+    FilterParameters.AcquireForSectionSynchronization.PageProtection = Reserved;
+    FilterParameters.AcquireForSectionSynchronization.SyncType = SyncType;
 
-        Status = FilterCallbacks->PreAcquireForSectionSynchronization(&CbData, &CompletionContext);
-        if (!NT_SUCCESS(Status))
-        {
-            FsRtlExitFileSystem();
-            return Status;
-        }
-
-        /* Should we do something in-between ? */
-
-        if (FilterCallbacks->PostAcquireForSectionSynchronization)
-        {
-            FilterCallbacks->PostAcquireForSectionSynchronization(&CbData, Status, CompletionContext);
-        }
-
-        /* Return here when the status is based on the synchonization type and write access to the file */
-        if (Status == STATUS_FSFILTER_OP_COMPLETED_SUCCESSFULLY ||
-            Status == STATUS_FILE_LOCKED_WITH_ONLY_READERS ||
-            Status == STATUS_FILE_LOCKED_WITH_WRITERS)
-        {
-            return Status;
-        }
+    if (FilterCallbacks)
+    {
+        /* Call pre callback */
+        Status = FsRtlpCallPreCallback(DeviceObject,
+                                       FileObject,
+                                       FS_FILTER_ACQUIRE_FOR_SECTION_SYNCHRONIZATION,
+                                       FilterCallbacks,
+                                       &FilterParameters,
+                                       FilterCallbacks->PreAcquireForSectionSynchronization,
+                                       &CompletionContext);
+        if (!NT_SUCCESS(Status)) return Status;
+        else goto End;
```

Comment:
I don't think that this is correct. It means that if a filter driver installs this callback, the file-system driver would never receive anything. If `FsRtlpCallPreCallback` was changed to call lower level filters as well, the file system driver would at least get this callback. A legacy FS driver that uses `FastDispatch->AcquireFileForNtCreateSection` would no longer work though, or even worse: it would "work", but synchronization would be broken.